h2o-algos/src/main/java/hex/gam/GAM.java

```diff
@@ -284,15 +280,98 @@ private void validateGamParameters() {
     }
     if (_parms._link == null || _parms._link.equals(GLMParameters.Link.family_default))
       _parms._link = _parms._family.defaultLink;
-
-
+    
     if ((_parms._family == GLMParameters.Family.multinomial || _parms._family == GLMParameters.Family.ordinal ||
             _parms._family == GLMParameters.Family.binomial)
             && response().get_type() != Vec.T_CAT) {
       error("_response_column", String.format("For given response family '%s', please provide a categorical" +
               " response column. Current response column type is '%s'.", _parms._family, response().get_type_str()));
     }
   }
+  
+  // verify and check thin plate regression smoothers specific parameters
+  public void checkThinPlateParams() {
+    if (_thinPlateSmoothersWithKnotsNum ==0)
+      return;
+    
+    _parms._num_knots_tp = new int[_thinPlateSmoothersWithKnotsNum];
+    System.arraycopy(_parms._num_knots_sorted, _cubicSplineNum, _parms._num_knots_tp, 0,
+            _thinPlateSmoothersWithKnotsNum);
+    int tpIndex = 0;
+    for (int index = 0; index < _parms._gam_columns.length; index++) {
+      if (_parms._bs_sorted[index] == 1) {
+        if (_parms._num_knots_sorted[index] < _parms._M[tpIndex] + 1) {
+          error("num_knots", "num_knots for gam column start with  " + _parms._gam_columns_sorted[index][0] +
+                  " did not specify enough num_knots.  It should be equal or greater than " + (_parms._M[tpIndex] + 1) + ".");
+        }
+        tpIndex++;
+      }
+    }
+  }
+  
+  // set default num_knots to 10 for gam_columns where there is no knot_id specified for CS smoothers
+  // for TP smoothers, default is set to be max of 10 or _M+2.
+  public void checkOrChooseNumKnots() {
+    if (_parms._num_knots == null)
+      _parms._num_knots = new int[_parms._gam_columns.length];  // different columns may have different
+    int tpCount = 0;
+    for (int index = 0; index < _parms._num_knots.length; index++) {  // set zero value _num_knots
+      if (_parms._knot_ids == null || (_parms._knot_ids != null && _parms._knot_ids[index] == null)) {  // knots are not specified
+        int numKnots = _parms._num_knots[index];
+        int naSum = 0;
+        for (int innerIndex = 0; innerIndex < _parms._gam_columns[index].length; innerIndex++) {
+          naSum += _parms.train().vec(_parms._gam_columns[index][innerIndex]).naCnt();
+        }
+        long eligibleRows = _train.numRows()-naSum;
+        if (_parms._num_knots[index] == 0) {  // set num_knots to default
+          int defaultRows = 10;
+          if (_parms._bs[index] == 1) {
+            defaultRows = Math.max(defaultRows, _parms._M[tpCount] + 2);
+            tpCount++;
+          }
+          _parms._num_knots[index] = eligibleRows < defaultRows ? (int) eligibleRows : defaultRows;
```

Comment:
Are you sure the `eliglibleRows` can't be larger than `int`?

Reply:
The algorithm will be very slow if num_knot is too big.  Hence, I am restricting it to be integer.

Reply:
wouldn't then be better to define `eliglibleRows` directly as integer?

Reply:
It depends.  If the number of rows in the dataframe is larger than an integer, there can be a problem doing

int eligibleRows = (int) frame.nrows().

Reply:
i see, then it's ok